npm/networkPolicyController_test.go

```diff
@@ -305,13 +315,54 @@ func TestDeleteNetworkPolicy(t *testing.T) {
 	defer close(stopCh)
 	f.newNetPolController(stopCh)
 
-	deleteNetPol(t, f, netPolObj)
+	deleteNetPol(t, f, netPolObj, DeletedFinalStateknownObject)
 	testCases := []expectedNetPolValues{
 		{1, 0, 0, false, true, 0, nil, 1, nil},
 	}
 	checkNetPolTestResult("TestDelNetPol", f, testCases)
 }
 
+func TestDeleteNetworkPolicyWithTombstone(t *testing.T) {
+	netPolObj := createNetPol()
+
+	f := newNetPolFixture(t)
+	f.isEnqueueEventIntoWorkQueue = false
+	f.netPolLister = append(f.netPolLister, netPolObj)
+	f.kubeobjects = append(f.kubeobjects, netPolObj)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	f.newNetPolController(stopCh)
+
+	netPolKey := getKey(netPolObj, t)
+	tombstone := cache.DeletedFinalStateUnknown{
+		Key: netPolKey,
+		Obj: netPolObj,
+	}
+
+	f.netPolController.deleteNetworkPolicy(tombstone)
+	testCases := []expectedNetPolValues{
+		{1, 0, 0, false, false, 0, nil, 0, nil},
+	}
+	checkNetPolTestResult("TestDeleteNetworkPolicyWithTombstone", f, testCases)
+}
+
+func TestDeleteNetworkPolicyWithTombstoneAfterAddingNetworkPolicy(t *testing.T) {
```

Comment:
<Nit> I feel like there is no major difference in functionality between TestDeleteNetworkPolicyWithTombstone and TestDeleteNetworkPolicyWithTombstoneAfterAddingNetworkPolicy as AddingNS does not care about tombstone and in deleting the way we contruct tombstone is same. So we can reduce the complexity these test files by having only one tescase and not send delefinalstate in deleteNetPol(t *testing.T, ...) func bu

Reply:
They are different.  First test case only check "delete" event function in each controller.
Second test case is to try to mimic the realistic situation which walks through "delete" event -> workqueue -> reconcile. Thus, second test case first adds resources and then we emulate the realistic story (i.e., missing event and relist happens). Then, it proceeds all required steps like cleaning up the applied states from "delete" event.
If we want to keep one test case, I would like to keep second one which has more comprehensive test.

I am not sure this increase complexity. It has one extra boolean variable for only delete event in UT and later we can do more sophisticated tests. But, if there is better way to deal with this, let me know.

Reply:
They are different. 
First test case only check "delete" event function in each controller. 
Second test case is to try to mimic the realistic situation which walks through "delete" event -> workqueue -> reconcile. Thus, second test case first adds resources and then we emulate the realistic story (i.e., missing event and relist happens). Then, it proceeds all required steps like cleaning up the applied states from "delete" event. 
If we want to keep one test case, I would like to keep second one which has more comprehensive test.

I am not sure this increase complexity. It has one extra boolean variable for only delete event in UT and later we can do more sophisticated tests. But, if there is better way to deal with this, let me know.